packages/network-enablement-controller/src/NetworkEnablementController.ts

```diff
@@ -0,0 +1,353 @@
+const metadata = {
+  enabledNetworkMap: {
+    persist: true,
+    anonymous: true,
+  },
+};
+
+const mergeEnabledMaps = (
+  base: EnabledMap,
+  override?: Partial<EnabledMap>,
+): EnabledMap => {
+  if (!override) {
+    return base;
+  }
+  return (Object.keys(base) as CaipNamespace[]).reduce((acc, ns) => {
+    acc[ns] = { ...base[ns], ...(override[ns] ?? {}) };
+    return acc;
+  }, {} as EnabledMap);
+};
+
+export class NetworkEnablementController extends BaseController<
+  typeof controllerName,
+  NetworkEnablementControllerState,
+  NetworkEnablementControllerMessenger
+> {
+  /**
+   * Creates a NetworkEnablementController instance.
+   *
+   * @param args - The arguments to this function.
+   * @param args.messenger - Messenger used to communicate with BaseV2 controller.
+   * @param args.state - Initial state to set on this controller.
+   */
+  constructor({
+    messenger,
+    state,
+  }: {
+    messenger: NetworkEnablementControllerMessenger;
+    state?: Partial<NetworkEnablementControllerState>;
+  }) {
+    // Call the constructor of BaseControllerV2
+    super({
+      messenger,
+      metadata,
+      name: controllerName,
+      state: {
+        enabledNetworkMap: mergeEnabledMaps(
+          defaultState().enabledNetworkMap,
+          state?.enabledNetworkMap,
+        ),
+      },
+    });
+
+    this.messagingSystem = messenger;
+
+    messenger.subscribe('NetworkController:networkAdded', ({ chainId }) => {
+      this.#ensureNetworkEntry(chainId, false);
+      this.#toggleNetwork(chainId, true);
+    });
+
+    messenger.subscribe('NetworkController:networkRemoved', ({ chainId }) => {
+      this.#removeNetworkEntry(chainId);
+    });
+  }
+
+  setEnabledNetwork(chainId: Hex | CaipChainId): void {
+    this.#toggleNetwork(chainId, true);
+  }
+
+  setDisabledNetwork(chainId: Hex | CaipChainId): void {
+    this.#toggleNetwork(chainId, false);
+  }
+
+  isNetworkEnabled(chainId: Hex | CaipChainId): boolean {
+    try {
+      const { namespace, storageKey } = this.#deriveKeys(chainId);
+      return Boolean(this.state.enabledNetworkMap[namespace]?.[storageKey]);
+    } catch {
+      return false;
+    }
+  }
+
+  getEnabledNetworksForNamespace(namespace: CaipNamespace): string[] {
+    return Object.entries(this.state.enabledNetworkMap[namespace] ?? {})
+      .filter(([, enabled]) => enabled)
+      .map(([id]) => id);
+  }
+
+  getAllEnabledNetworks(): Record<CaipNamespace, string[]> {
+    return (
+      Object.keys(this.state.enabledNetworkMap) as CaipNamespace[]
+    ).reduce(
+      (acc, ns) => {
+        acc[ns] = this.getEnabledNetworksForNamespace(ns);
+        return acc;
+      },
+      {} as Record<CaipNamespace, string[]>,
+    );
+  }
+
+  // ---------------------------- Internals ----------------------------------
+
+  #deriveKeys(chainId: Hex | CaipChainId) {
+    const caipId: CaipChainId = isHexString(chainId)
+      ? toEvmCaipChainId(chainId as Hex)
+      : (chainId as CaipChainId);
```

Comment:
The type assertion shouldn't be necessary here. If you flip this around and use `isCaipChainId` from `@metamask/utils` you can avoid it:

``` suggestion
    const caipId: CaipChainId = isCaipChainId(chainId)
      ? chainId
      : toEvmCaipChainId(chainId);
```